### PR TITLE
[serve] Clean up API reference and various docstrings

### PR DESCRIPTION
**Pull request by edoakes**
## Why are these changes needed?

Cleans up the API reference page formatting, adds/updates docstrings where applicable (corrects typing + verbiage, adds inline examples).

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/serve/api.py [hunk 1]
```diff
@@ -296,62 +282,44 @@ def deployment(
...
+            If not provided, the name of the class or function will be used.
+        num_replicas: The number of replicas to run that will handle requests to
+            this deployment. Defaults to 1.
+        init_args: [DEPRECATED] These should be passed to `.bind()` instead.
+        init_kwargs: [DEPRECATED] These should be passed to `.bind()` instead.
+        route_prefix: Requests to paths under this HTTP path prefix will be routed
+            to this deployment. Defaults to '/{name}'. This can only be set for the
+            ingress (top-level) deployment of an application.
+        ray_actor_options: Options to be passed to the Ray actor decorator such as
+            resource requirements. Valid options are `accelerator_type`, `memory`,
```
Comment (akshay-anyscale) [inline python/ray/serve/api.py:310]: I think this has some of the configs left out like autoscaling config, healthcheck, graceful shutdown

**Comment by ollie-iterators** [comment]
The errors seem to be unrelated to this PR. The doc test error "ModuleNotFoundError" should be fixed by merging in recent changes to the main branch.

File: python/ray/serve/api.py [hunk 2]
```diff
@@ -296,62 +282,44 @@ def deployment(
...
+            `num_cpus`, `num_gpus`, `object_store_memory`, `resources`,
+            and `runtime_env`.
+        user_config: Config to pass to the reconfigure method of the deployment. This
+            can be updated dynamically without restarting the replicas of the
+             deployment. The user_config must be fully JSON-serializable.
```
Comment (zcin) [inline python/ray/serve/api.py:315]: I think something weird is happening here:

Screen Shot 2023-05-01 at 8 32 22 AM

File: python/ray/serve/handle.py [hunk 3]
```diff
@@ -78,8 +78,9 @@ class HandleOptions:
 class RayServeHandle:
     """A handle used to make requests from one deployment to another.
 
-    This is used to compose multiple deployments in a single application by binding
-    them together when building the application. For example:
+    This is used to compose multiple deployments in a single application. When building
+    the application, deployments passed as arguments to others via `.bind()` will be
+    subsitututed for this handle at runtime. For example:
```
Comment (zcin) [inline python/ray/serve/handle.py:83]:
```suggestion
    This is used to compose multiple deployments in a single application. After building
    the application, this handle will be substituted at runtime for deployments passed 
    as arguments via `.bind()`. For example:
```

File: python/ray/serve/deployment.py [hunk 4]
```diff
@@ -67,6 +95,32 @@ def __getattr__(self, name: str) -> Any:
 
 @PublicAPI
 class Deployment:
+    """Defines group of Ray actors to be deployed in an `Application`.
+
+    Deployments are used to define a Serve application by binding them together with
+    arguments (and possibly other deployments for composition).
+
+    The resulting `Application` can be deployed using `serve.run` or via a Serve config
+    file.
```
Comment (zcin) [inline python/ray/serve/deployment.py:104]: The summary line "Defines group of Ray actors to be deployed in an `Application`." feels a bit confusing to me. 

Also, the description is talking a lot about applications - would it be better to describe more concretely what a deployment is, e.g. "created by using the @serve.deployment decorator" or "can either be a function or class"? Maybe mention at the very end that the deployments can be composed together to form an application?

**Review by zcin — COMMENTED** [review]
Looks good overall! A few nits

File: python/ray/serve/api.py (hunk 2 above)
Comment (edoakes) [reply python/ray/serve/api.py:315]: there was an errant space :) fixed

File: python/ray/serve/api.py (hunk 1 above)
Comment (edoakes) [reply python/ray/serve/api.py:312]: added!

File: python/ray/serve/handle.py (hunk 3 above)
Comment (edoakes) [reply python/ray/serve/handle.py:83]: nice

**Comment by edoakes** [comment]
@zcin @akshay-anyscale comments addressed

**Review by akshay-anyscale — APPROVED** [review]

File: python/ray/serve/deployment.py [hunk 5]
```diff
@@ -67,6 +95,32 @@ def __getattr__(self, name: str) -> Any:
 
 @PublicAPI
 class Deployment:
+    """Class (or function) decorated with the `@serve.deployment` decorator.
+
+    This class will be run by a number of actor replicas and called to handle requests
+    to those replicas.
```
Comment (zcin) [inline python/ray/serve/deployment.py:101]:
```suggestion
    This class will be run on a number of replica actors. Requests to those replicas
    will call this class.
```

**Review by zcin — APPROVED** [review]
LGTM!

**Comment by edoakes** [comment]
@ray-project/ray-docs can I get a review on this please

File: python/ray/serve/api.py [hunk 6]
```diff
@@ -296,62 +282,56 @@ def deployment(
...
+
+        @serve.deployment(num_replicas=2)
+        class MyDeployment:
+            pass
+
+        app = MyDeployment.bind()
+
+    Args:
+        name: Name uniquely identifying this deployment within the application.
+            If not provided, the name of the class or function will be used.
```
Comment (angelinalg) [inline python/ray/serve/api.py:301]:
```suggestion
            If not provided, the name of the class or function is be used.
```

File: python/ray/serve/api.py [hunk 7]
```diff
@@ -296,62 +282,56 @@ def deployment(
...
+        num_replicas: The number of replicas to run that will handle requests to
```
Comment (angelinalg) [inline python/ray/serve/api.py:302]:
```suggestion
        num_replicas: The number of replicas to run that handles requests to
```

File: python/ray/serve/api.py [hunk 8]
```diff
@@ -296,62 +282,56 @@ def deployment(
...
+            this deployment. Defaults to 1.
+        autoscaling_config: Parameters to configure autoscaling behavior. If this
+            is set, `num_replicas` cannot be set.
+        init_args: [DEPRECATED] These should be passed to `.bind()` instead.
+        init_kwargs: [DEPRECATED] These should be passed to `.bind()` instead.
+        route_prefix: Requests to paths under this HTTP path prefix will be routed
```
Comment (angelinalg) [inline python/ray/serve/api.py:308]:
```suggestion
        route_prefix: Requests to paths under this HTTP path prefix are routed
```

File: python/ray/serve/api.py [hunk 9]
```diff
@@ -296,62 +282,56 @@ def deployment(
...
+            to this deployment. Defaults to '/{name}'. This can only be set for the
+            ingress (top-level) deployment of an application.
+        ray_actor_options: Options to be passed to the Ray actor decorator such as
```
Comment (angelinalg) [inline python/ray/serve/api.py:311]:
```suggestion
        ray_actor_options: Options to be passed to the Ray actor decorator, such as
```

File: python/ray/serve/api.py [hunk 10]
```diff
@@ -296,62 +282,56 @@ def deployment(
...
+            resource requirements. Valid options are `accelerator_type`, `memory`,
+            `num_cpus`, `num_gpus`, `object_store_memory`, `resources`,
+            and `runtime_env`.
+        user_config: Config to pass to the reconfigure method of the deployment. This
+            can be updated dynamically without restarting the replicas of the
+            deployment. The user_config must be fully JSON-serializable.
+        max_concurrent_queries: The maximum number of queries that will be sent to a
```
Comment (angelinalg) [inline python/ray/serve/api.py:318]:
```suggestion
        max_concurrent_queries: The maximum number of queries that are sent to a
```

File: python/ray/serve/api.py [hunk 11]
```diff
@@ -296,62 +282,56 @@ def deployment(
...
+            replica of this deployment without receiving a response. Defaults to 100.
+        health_check_period_s: How often the health check will be called on the replica.
```
Comment (angelinalg) [inline python/ray/serve/api.py:320]:
```suggestion
        health_check_period_s: How often the health check is called on the replica.
```

File: python/ray/serve/api.py [hunk 12]
```diff
@@ -296,62 +282,56 @@ def deployment(
...
+            Defaults to 10s. The health check is by default a no-op actor call to the
+            replica, but you can define your own as a "check_health" method that raises
+            an exception when unhealthy.
+        health_check_timeout_s: How long to wait for a health check method to return
+            before considering it failed. Defaults to 30s.
+        graceful_shutdown_wait_loop_s: Duration that replicas will wait until there is
```
Comment (angelinalg) [inline python/ray/serve/api.py:326]:
```suggestion
        graceful_shutdown_wait_loop_s: Duration that replicas wait until there is
```

File: python/ray/serve/api.py [hunk 13]
```diff
@@ -296,62 +282,56 @@ def deployment(
...
+            no more work to be done before shutting down.
+        graceful_shutdown_timeout_s: Duration that a replica can be gracefully shutting
+            down before being forcefully killed.
+        is_driver_deployment: [EXPERIMENTAL] when set, exactly one
+            replica of this deployment will run on every node (like a daemon set).
```
Comment (angelinalg) [inline python/ray/serve/api.py:331]:
```suggestion
            replica of this deployment runs on every node (like a daemon set).
```

File: python/ray/serve/api.py [hunk 14]
```diff
@@ -466,12 +446,16 @@ def run(
...
+
+        handle = serve.run(MyDeployment.bind())
+        ray.get(handle.remote())
 
     Args:
-        target (Union[Application, BuiltApplication]):
-            A Serve application returned from `Deployment.bind()` or a built application
-            returned from `serve.build()`.
+        target:
+            A Serve application returned from `Deployment.bind()`.
```
Comment (angelinalg) [inline python/ray/serve/api.py:458]:
```suggestion
            A Serve application returned by `Deployment.bind()`.
```

File: python/ray/serve/handle.py [hunk 15]
```diff
@@ -78,8 +78,11 @@ class HandleOptions:
 class RayServeHandle:
     """A handle used to make requests from one deployment to another.
 
-    This is used to compose multiple deployments in a single application by binding
-    them together when building the application. For example:
+    This is used to compose multiple deployments in a single application. After building
+    the application, this handle will be substituted at runtime for deployments passed
```
Comment (angelinalg) [inline python/ray/serve/handle.py:82]:
```suggestion
    the application, this handle is substituted at runtime for deployments passed
```

**Review by richardliaw — APPROVED** [review]

File: python/ray/serve/api.py [hunk 16]
```diff
@@ -129,21 +128,28 @@ def shutdown() -> None:
...
         RayServeException: if not called from within a Ray Serve deployment.
 
     Example:
-        >>> from ray import serve
-        >>> # deployment_name
-        >>> serve.get_replica_context().deployment # doctest: +SKIP
-        >>> # deployment_name#krcwoa
-        >>> serve.get_replica_context().replica_tag # doctest: +SKIP
+
+        .. code-block:: python
```
Comment (richardliaw) [inline python/ray/serve/api.py:141]: eventually we will move off this code-block thing, fyi

File: python/ray/serve/deployment.py [hunk 17]
```diff
@@ -32,18 +32,46 @@
...
-@PublicAPI
+@PublicAPI(stability="beta")
 class Application(DAGNodeBase):
-    """Returned from `Deployment.bind()`.
+    """One or more deployments bound with arguments that can be deployed together.
 
     Can be passed into another `Deployment.bind()` to compose multiple deployments in a
     single application, passed to `serve.run`, or deployed via a Serve config file.
+
+    For example, to define an app and run it in Python:
```
Comment (richardliaw) [inline python/ray/serve/deployment.py:42]: define an Application (shortened as "app")

**Comment by angelinalg** [comment]
Had some minor copy edits. Please use the `Add suggestion to batch` button if you want to batch the suggestions into a single commit. Hope that helps.

File: doc/source/serve/api/index.md [hunk 18]
```diff
@@ -1,12 +1,395 @@
...
+
+```{eval-rst}
+.. module:: ray
+```
+
+### Writing Applications
+
+<!---
+NOTE: `serve.deployment` and `serve.Deployment` have an autosummary-generated filename collision due to case insensitivity.
+This is fixed by added custom filename mappings in `source/conf.py` (look for "autosummary_filename_map").
```
[the remark this answers is not in the document]
Comment (edoakes) [reply doc/source/serve/api/index.md:16]: I think it's actually correct as-is?

**Review by angelinalg — APPROVED** [review]
LGTM

File: python/ray/serve/api.py (hunk 16 above)
Comment (edoakes) [reply python/ray/serve/api.py:141]: onto what?

**Pull request merged**